Support .d.ts files

kitsonk opened the pull request:
Fixes #1432

I have a MVP where the following is working:

```ts
/// <deno-types path="./foo.d.ts" />
import * as foo from "./foo.js";

/// <deno-types path="./bar.d.ts" />
import * as bar from "./bar.js";

console.log(foo.foo, bar.bar);
```

Where instead of TypeScript trying to load `foo.js` and `bar.js`, it instead loads `foo.d.ts` and `bar.d.ts`.  There are quite a few more scenarios that need to be supported and I need to try it out with some real world code, but it looks like it will work.

ry commented:
Just a quick atheistic comment. 
```
/// <deno-types path="./bar.d.ts" />
```
I actually preferred your original `//# typeDefinition="./bar.d.ts"` but suggested triple-slash reference directive as a way of potentially being able to extract the information from the typescript AST. If `/// <reference ...something.../>`  isn't the syntax we use, might as well use the `//#` syntax.

Shortening `//# typeDefinition=` to `//# types=` could be done without introducing confusion, I think.

ry commented on tests/type_definitions/main.ts:
```diff
@@ -0,0 +1,7 @@
+import * as foo from "./foo.js";
+
+/// <deno-types path="./bar.d.ts" />
+import * as bar from "./bar.js";
```
Wouldn't it be nicer to be able to declare this in bar.js ?

kitsonk replied:
That assumes you have control of `bar.js` which I didn't think would be a common workload.  Also, in order for the compiler to see that it has to resolve and load the `.js` file.  In this case, the `.js` file is never loaded by the compiler, the `.d.ts` is instead to give TypeScript the types.

The typical use case is going to be, in my opinion, "I am using an external library named "foo", that library has types at `@types/foo`, I want to use that library in Deno, but I also want that code to work other places".  I will write something like this, likely in something like `deps.ts`:

```js
//# types=https://unpkg.com/@types/foo@~1.0.0/index.d.ts
import * as foo from "https://unpkg.com/foo@~1.0.0/index.js";
```

If you have mixed mode Deno project though, you don't need this feature, as `allowJs` is on by default, which allows you to resolve `.js` files and import them into Deno, and then `checkJs` would allow you to type check the files based on the JavaScript itself, plus and JSDoc comments.  (I also believe `// @ts-check` in each JavaScript file would also work, though I haven't tested it).

ry replied:
> That assumes you have control of bar.js

Right. I see now.

In typescript currently this is done with the `<reference type="blah"/>` directive? The `//#` syntax now looks strange since it's not applying to the current file.

kitsonk replied:
Sort of...

From the TypeScript documentation:

> For example, including `/// <reference types="node" />` in a declaration file declares that this file uses names declared in `@types/node/index.d.ts;` and thus, this package needs to be included in the compilation along with the declaration file.

It is more of a directive for `.d.ts` files to say "I also need these types" and relies upon the "magical" resolution within `node_modules`.  We might actually have a bit of a problem with that now that you mention it, but at least this would be a step forward and we can figure out how to address that in the future.

I agree it does look funny, the only other suggestion, is that there are a couple TypeScript compiler "hints" like `@ts-check` and `@ts-nocheck` and `@ts-ignore`.  A "hint" style syntax might work which could be something like:

```ts
// @deno-types="https://unpkg.com/@types/foo@~1.0.0/index.d.ts"
import * as foo from "https://unpkg.com/foo@~1.0.0/index.js";
```

kitsonk commented:
Ok, so my working branch, this now works:

```ts
// @deno-types="https://unpkg.com/dayjs@1.8.15/index.d.ts"
import dayjs from "https://unpkg.com/dayjs@1.8.15/esm/index.js";

const date = dayjs();
console.log(date.seconds());
```

Which has an error (the API is `second()`, not `seconds()`), so it now outputs:

```
Compile file:///Users/kkelly/github/deno/tests/type_definitions/remote/main.ts
Download https://unpkg.com/dayjs@1.8.15/index.d.ts
error TS2551: Property 'seconds' does not exist on type 'Dayjs'. Did you mean 'second'?

► file:///Users/kkelly/github/deno/tests/type_definitions/remote/main.ts:5:18

5 console.log(date.seconds());
                   ~~~~~~~

  'second' is declared here.

    ► https://unpkg.com/dayjs@1.8.15/index.d.ts:46:5

    46     second(): number
           ~~~~~~

```

Fixing it then outputs:

```
Compile file:///Users/kkelly/github/deno/tests/type_definitions/remote/main.ts
Download https://unpkg.com/dayjs@1.8.15/index.d.ts
Download https://unpkg.com/dayjs@1.8.15/esm/index.js
Download https://unpkg.com/dayjs@1.8.15/esm/constant
Download https://unpkg.com/dayjs@1.8.15/esm/utils
Download https://unpkg.com/dayjs@1.8.15/esm/locale/en
Download https://unpkg.com/dayjs@1.8.15/esm/constant.js
Download https://unpkg.com/dayjs@1.8.15/esm/utils.js
Download https://unpkg.com/dayjs@1.8.15/esm/locale/en.js
32
```

Getting excited...  The other part I need to do though is that if the value of the directive looks like a directory to resolve the `.d.ts` files relative to that directory, in order to support UMD type definitions (where each module has its own `.d.ts` file).

kitsonk commented:
Ok, I have tested now with lodash-es but using the types from `@types/lodash` and that appears to be all working:

```ts
// @deno-types="https://unpkg.com/@types/lodash@4.14.136/index.d.ts"
import * as _ from "https://unpkg.com/lodash-es@4.17.15/lodash.js";

console.log(_.add(1, 2));
```

The change I made is that the hint only applies to the next `import`/`export` statement.  My original thought would be that we might have several imports in a row and that we would resolve the type definition for each on based on a path, but that doesn't make a lot of sense in practice.

We still won't be able to support all type definitions at the moment, since some try to import or reference types from other type libraries.  (e.g. `/// <reference types="node" />`).  Because we don't magically resolve them, we don't know where to look.  Right now they simply get passed upstream and Rust complains about not being able to resolve the module, but provides a cryptic message.  I want to catch those earlier and have something more meaningful.

ry commented:
Are you able to load the TS compiler with this? (Review soon...)

kitsonk commented:
> Are you able to load the TS compiler with this?

That is my next use case to test.

kitsonk commented:
@ry so it works from this PRs perspective.  The following:

Outputs:

```
$ deno --reload tests/type_definitions/typescript/main.ts
Compile file:///Users/kkelly/github/deno/tests/type_definitions/typescript/main.ts
Download https://unpkg.com/typescript@3.5.3/lib/typescript.d.ts
error TS2551: Property 'versions' does not exist on type 'typeof ts'. Did you mean 'version'?

► file:///Users/kkelly/github/deno/tests/type_definitions/typescript/main.ts:4:16

4 console.log(ts.versions);
                 ~~~~~~~~

  'version' is declared here.

    ► https://unpkg.com/typescript@3.5.3/lib/typescript.d.ts:19:11

    19     const version: string;
                 ~~~~~~~

```

But the way the `typescript.js` is formed, nothing gets exported as the module, so Deno loads it, but we get an empty object returned back.  Looking at the script, I am a little confused how it works at all, under node or rollup.  It is a global script that creates a `var ts` variable (plus a couple others) and runs it through a series of IIFE.  I am not sure how it "exports" anything. 😕

afinch7 commented:
If you look at the end of https://raw.githubusercontent.com/microsoft/TypeScript/master/lib/typescript.js
You can see this commonjs export
```
    if (typeof module !== "undefined" && module.exports) {
        module.exports = ts;
    }
```
not sure if this is the only export. I did manage to get typescript imported with .d.ts types after creating a rollup bundle that just reexports everything.

kitsonk commented:
@afinch7 that makes sense, and I don't know how I missed it.  Currently we don't have an easy way to load stuff like this in Deno (nor do I think we should).  With known things like TypeScript, it should be fairly easy to wrap it and have it be a ESM file.  But in theory it is tangental to this PR.  So we can now load the types, but we can't load the module.

Since really TypeScript should support output to ESM, I opened issue microsoft/TypeScript#32949 to request it.

ry commented on website/manual.md:
```diff
@@ -580,6 +580,48 @@ import { test, assertEquals } from "./deps.ts";
+
+The TypeScript compiler supports triple-slash directives, including a type
+reference directive. If Deno used this, it would interfere with the behavior of
+the TypeScript compiler.
```
We could maybe intercept and remove a triple slash reference...

I'm still not sure about this syntax.

kitsonk replied:
If we used something that conflicted with TypeScript, we could remove it for Deno, but we would want it to be a noop everywhere else, otherwise we would have Deno only TypeScript.

Even for Deno, if it overlapped TypeScript syntax, we would need a way to disambiguate it, which I don't think would be easy, so we could use a triple-slash directive, but it would have to be something TypeScript currently ignores, so the following would be out:

```ts
/// <reference path="..." />
/// <reference types="..." />
/// <reference lib="..." />
/// <reference no-default-lib="true"/>
```

So something like this "could" work and would be terse enough:

```ts
/// <reference deno="..." />
```

The biggest "challenge" I see is that these are like include statements, and they effect the file they are in.  What we need to support in Deno is something where certain statements are impacted, so they are external and their position matters.  Because there is such a dramatic behaviour change, it maybe confusing.

ry commented:
This patch seems to have reduced the thread count for cold_relative_import from 13 to 11. That pleasant but surprising.

<img width="688" alt="Screen Shot 2019-08-22 at 12 29 00 PM" src="https://user-images.githubusercontent.com/80/63532354-8089b800-c4d8-11e9-9df9-0e531a7514a3.png">

kitsonk commented:
Thanks. Yeah I think it is best to iterate on it, even if we change the syntax. Hard to tell what we don't like. It may also struggle with some real world use cases which we can start to address.

I implemented a straight forward cache in the compiler that existed previously. I did it to help ensure once we had replaced a file we knew it. I suspect that had a side impact of reducing the threads that went back to Rust.
